### ISSUE TITLE
Panic when trying to get a mutable reference on Request RC

### ISSUE THREAD

**Issue by Sytten** [actix/actix-web#2508]
## Expected Behavior

Trying to upgrade from actix-web v4 beta 10 to beta 14.
It should work correctly and no panic. It was working fine with beta 10.

## Current Behavior
```
thread 'actix-rt|system:0|arbiter:1' panicked at 'called `Option::unwrap()` on a `None` value', /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-web-4.0.0-beta.14/src/request.rs:140:43
```

Backtrace:
```
   0: rust_begin_unwind
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panicking.rs:101:14
   2: core::panicking::panic
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panicking.rs:50:5
   3: core::option::Option<T>::unwrap
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/option.rs:735:21
   4: actix_web::request::HttpRequest::match_info_mut
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-web-4.0.0-beta.14/src/request.rs:140:14
   5: actix_web::service::ServiceRequest::match_info_mut
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-web-4.0.0-beta.14/src/service.rs:226:9
   6: <actix_web::service::ServiceRequest as actix_router::Resource<actix_router::url::Url>>::resource_path
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-web-4.0.0-beta.14/src/service.rs:291:9
   7: actix_router::resource::ResourceDef::capture_match_info_fn
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-router-0.5.0-beta.2/src/resource.rs:701:20
   8: actix_router::router::Router<T,U>::recognize_fn
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-router-0.5.0-beta.2/src/router.rs:69:16
   9: <actix_web::app_service::AppRouting as actix_service::Service<actix_web::service::ServiceRequest>>::call
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-web-4.0.0-beta.14/src/app_service.rs:304:19
  10: <actix_cors::middleware::CorsMiddleware<S> as actix_service::Service<actix_web::service::ServiceRequest>>::call
             at /Users/Sytten/.cargo/git/checkouts/actix-extras-b20316b0f9e556d8/74ec115/actix-cors/src/middleware.rs:171:23
  11: <actix_web::app_service::AppInitService<T,B> as actix_service::Service<actix_http::request::Request>>::call
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-web-4.0.0-beta.14/src/app_service.rs:223:9
  12: actix_http::h1::dispatcher::InnerDispatcher<T,S,B,X,U>::handle_request
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-http-3.0.0-beta.15/src/h1/dispatcher.rs:509:24
  13: actix_http::h1::dispatcher::InnerDispatcher<T,S,B,X,U>::poll_request
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-http-3.0.0-beta.15/src/h1/dispatcher.rs:623:33
  14: <actix_http::h1::dispatcher::Dispatcher<T,S,B,X,U> as core::future::future::Future>::poll
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-http-3.0.0-beta.15/src/h1/dispatcher.rs:921:21
  15: <actix_http::service::HttpServiceHandlerResponse<T,S,B,X,U> as core::future::future::Future>::poll
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-http-3.0.0-beta.15/src/service.rs:620:45
  16: <actix_service::and_then::AndThenServiceResponse<A,B,Req> as core::future::future::Future>::poll
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-service-2.0.0/src/and_then.rs:114:37
  17: <actix_service::and_then::AndThenServiceResponse<A,B,Req> as core::future::future::Future>::poll
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-service-2.0.0/src/and_then.rs:112:17
  18: engine::api::Api::handle::{{closure}}
             at ./src/api/mod.rs:144:17
  19: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/future/mod.rs:80:19
  20: engine::server::dispatcher::Dispatcher::dispatch::{{closure}}
             at ./src/server/dispatcher.rs:52:17
  21: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/future/mod.rs:80:19
  22: engine::server::Server::bind::{{closure}}::{{closure}}::{{closure}}
             at ./src/server/mod.rs:38:27
  23: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/future/mod.rs:80:19
  24: <actix_server::service::StreamService<S,I> as actix_service::Service<(actix_server::worker::WorkerCounterGuard,actix_server::socket::MioStream)>>::call::{{closure}}
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-server-2.0.0-rc.1/src/service.rs:70:29
  25: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/future/mod.rs:80:19
  26: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/task/core.rs:147:17
  27: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/loom/std/unsafe_cell.rs:14:9
  28: tokio::runtime::task::core::CoreStage<T>::poll
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/task/core.rs:137:13
  29: tokio::runtime::task::harness::poll_future::{{closure}}
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/task/harness.rs:437:23
  30: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panic/unwind_safe.rs:271:9
  31: std::panicking::try::do_call
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/panicking.rs:403:40
  32: <unknown>
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/panicking.rs:431:6
  33: std::panicking::try
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/panicking.rs:367:19
  34: std::panic::catch_unwind
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/panic.rs:129:14
  35: tokio::runtime::task::harness::poll_future
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/task/harness.rs:424:19
  36: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/task/harness.rs:89:9
  37: tokio::runtime::task::harness::Harness<T,S>::poll
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/task/harness.rs:59:15
  38: tokio::runtime::task::raw::poll
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/task/raw.rs:113:5
  39: tokio::runtime::task::raw::RawTask::poll
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/task/raw.rs:70:18
  40: tokio::runtime::task::Notified<S>::run
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/task/mod.rs:166:9
  41: tokio::task::local::LocalSet::tick::{{closure}}
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/task/local.rs:532:54
  42: tokio::coop::with_budget::{{closure}}
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/coop.rs:106:9
  43: std::thread::local::LocalKey<T>::try_with
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/thread/local.rs:399:16
  44: std::thread::local::LocalKey<T>::with
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/thread/local.rs:375:9
  45: tokio::coop::with_budget
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/coop.rs:99:5
  46: tokio::coop::budget
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/coop.rs:76:5
  47: tokio::task::local::LocalSet::tick
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/task/local.rs:532:31
  48: <tokio::task::local::RunUntil<T> as core::future::future::Future>::poll::{{closure}}
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/task/local.rs:672:16
  49: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/macros/scoped_tls.rs:61:9
  50: tokio::task::local::LocalSet::with
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/task/local.rs:573:9
  51: <tokio::task::local::RunUntil<T> as core::future::future::Future>::poll
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/task/local.rs:658:9
  52: tokio::task::local::LocalSet::run_until::{{closure}}
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/task/local.rs:516:9
  53: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/future/mod.rs:80:19
  54: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/future/future.rs:119:9
  55: tokio::runtime::basic_scheduler::Inner<P>::block_on::{{closure}}::{{closure}}
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/basic_scheduler.rs:208:62
  56: tokio::coop::with_budget::{{closure}}
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/coop.rs:106:9
  57: std::thread::local::LocalKey<T>::try_with
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/thread/local.rs:399:16
  58: std::thread::local::LocalKey<T>::with
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/thread/local.rs:375:9
  59: tokio::coop::with_budget
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/coop.rs:99:5
  60: tokio::coop::budget
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/coop.rs:76:5
  61: tokio::runtime::basic_scheduler::Inner<P>::block_on::{{closure}}
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/basic_scheduler.rs:208:39
  62: tokio::runtime::basic_scheduler::enter::{{closure}}
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/basic_scheduler.rs:299:29
  63: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/macros/scoped_tls.rs:61:9
  64: tokio::runtime::basic_scheduler::enter
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/basic_scheduler.rs:299:5
  65: tokio::runtime::basic_scheduler::Inner<P>::block_on
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/basic_scheduler.rs:197:9
  66: tokio::runtime::basic_scheduler::InnerGuard<P>::block_on
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/basic_scheduler.rs:452:9
  67: tokio::runtime::basic_scheduler::BasicScheduler<P>::block_on
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/basic_scheduler.rs:157:24
  68: tokio::runtime::Runtime::block_on
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/mod.rs:450:46
  69: tokio::task::local::LocalSet::block_on
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/task/local.rs:477:9
  70: actix_rt::runtime::Runtime::block_on
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-rt-2.5.0/src/runtime.rs:80:9
  71: actix_rt::arbiter::Arbiter::with_tokio_rt::{{closure}}
             at /Users/Sytten/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-rt-2.5.0/src/arbiter.rs:145:21
```

We do some funky stuff because we have multiple services running on the same port (web service and a proxy):
```rust
let listener = TcpListener::bind(&addr.to_string()).map_err(...)?;

ServerBuilder::default().listen("dispatcher-service", listener, move || {
    actix_service::fn_service(move |stream: tokio::net::TcpStream| async {
        match Dispatcher::dispatch(stream).await {
            Ok(_) => Ok(()),
            Err(err) => {
                log::error!("[Dispatcher] {}", err);
                Err(err)
            }
        }
     })
})?.run();
```

We then peek in the stream and dispatch to the right place accordingly.
For the web interface it goes into `Api::handle` (which uses async-graphql):
```rust
 pub async fn handle(&self, stream: TcpStream) {
        let schema = Schema::build(
            QueryRoot::default(),
            MutationRoot::default(),
            SubscriptionRoot::default(),
        )
        .finish();

        let app = App::new()
            .app_data(Data::new(schema))
            .service(web::resource("/graphql").guard(guard::Post()).to(graphql))
            .route("/{filename:.*}", web::get().to(default))
            .default_service(web::route().to(default));

        let config = actix_service::map_config(app, move |_| AppConfig::default());
        let builder = actix_http::HttpService::build()
            .keep_alive(10)
            .client_timeout(10)
            .finish(config)
            .tcp();

        let service = builder
            .new_service(())
            .await
            .expect("Could not create web service.");

        let _ = service.call(stream).await;
    }
```

It fails after the `service.call`.

## Your Environment

* Rust Version (I.e, output of `rustc -V`): rustc 1.56.1 (59eed8a2a 2021-11-01)
* Actix Web Version: v4 beta 14

**Comment by sfisol** [comment]
I got the same issue, but I used
```actix-cors = { git = "https://github.com/actix/actix-extras.git", branch = "beta-13-test" }```
To be able to compile my project with `actix-4.0.0-beta.14`.

After it compiles and runs I got similar error:
```thread 'actix-rt|system:0|arbiter:0' panicked at 'called `Option::unwrap()` on a `None` value', /home/szaman/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-web-4.0.0-beta.14/src/request.rs:140:43```

When I removed cors from my project:
```            // .wrap(Cors::default())```
then everything works again.

Maybe try to comment out different middleware you're using to spot the problem on your side.

EDIT: It fails also with ```actix-cors = "0.6.0-beta.5"```

**Comment by Sytten** [comment]
Good point I also use `cors` but didnt include it in my comment above. So it is most likely a middleware problem.

**Comment by tamasfe** [comment]
I don't think it's a middleware issue, [capture_match_info_fn](https://github.com/actix/actix-web/blob/cea44be67059a5270a892566323e9055809676de/actix-router/src/resource.rs#L685) requires a mutable borrow to the request due to [this](https://github.com/actix/actix-web/blob/cea44be67059a5270a892566323e9055809676de/src/service.rs#L226). Thus any middleware that clones the request will break the assumption that the request can be mutated later during matching.

I don't know enough about actix-web to fix this, just sharing my findings.

**Comment by robjtede** [comment]
fixed with release of actix-cors `0.6.0-beta.6`
